File: solutions/python/dp-2d/unique-paths.py
# LeetCode 62 — Unique Paths (Medium)
# Category: 2-D Dynamic Programming · Approach: Rolling Row DP
# Time: O(m·n) | Space: O(n)
# Source: https://leetcode.com/problems/unique-paths/

def unique_paths(m: int, n: int) -> int:
    row = [1] * n
    for _ in range(1, m):
        for c in range(1, n):
            row[c] += row[c - 1]
    return row[-1]

File: solutions/typescript/dp-2d/unique-paths.ts
// LeetCode 62 — Unique Paths (Medium)
// Category: 2-D Dynamic Programming · Approach: Rolling Row DP
// Time: O(m·n) | Space: O(n)
// Source: https://leetcode.com/problems/unique-paths/

function uniquePaths(m: number, n: number): number {
  const row = new Array<number>(n).fill(1);
  for (let r = 1; r < m; r++) {
    for (let c = 1; c < n; c++) {
      row[c] += row[c - 1];
    }
  }
  return row[n - 1];
}
